mesonbuild/interpreterbase/decorators.py

```diff
@@ -505,18 +505,24 @@ def check_value_type(types_tuple: T.Tuple[T.Union[T.Type, ContainerTypeInfo], ..
         @wraps(f)
         def wrapper(*wrapped_args: T.Any, **wrapped_kwargs: T.Any) -> T.Any:
 
-            def emit_feature_change(values: T.Dict[str, T.Union[str, T.Tuple[str, str]]], feature: T.Union[T.Type['FeatureDeprecated'], T.Type['FeatureNew']]) -> None:
+            def emit_feature_change(values: T.Dict[_T, T.Union[str, T.Tuple[str, str]]], feature: T.Union[T.Type['FeatureDeprecated'], T.Type['FeatureNew']]) -> None:
                 for n, version in values.items():
-                    if isinstance(value, (dict, list)):
+                    warn = False
+                    if isinstance(version, tuple):
+                        version, msg = version
+                    else:
+                        msg = None
+
+                    if n in (dict, list):
```

Comment:
```suggestion
                    if n in {dict, list}:
```
always use a set for `in` checks when the types are hashable, it's faster